sql/sql_acl.cc

```diff
@@ -10512,7 +10595,10 @@ int mysql_alter_user(THD* thd, List<LEX_USER> &users_list)
   {
     LEX_USER* lex_user= get_current_user(thd, tmp_lex_user, false);
     if (!lex_user || replace_user_table(thd, tables.user_table(), lex_user, 0,
-                                        false, false, true))
+                                        false, false, true) ||
+        (update_account_locking &&
+         lock_user_account(thd, tables.user_table(), lex_user,
```

Comment:
There is a small problem with this approach. In practice it will matter very little, but it can cause the following problem:

replace_user_table succeeds.
lock_user_acount fails due to say IO error.

The current status of the user table has an account that is not locked, yet it exists, with privileges possibly changed. This means that one can actually log in with the user and potentially run queries the admin never intended to allow.

Quick suggestion how to fix this: (yes, I know, this replace_xxx function is begging for a refactor)
Move the call of lock_user_account inside replace_user_table, make use of thd->lex to fetch the required arguments and run the logic before a call to the handlerton's write method.